### ISSUE TITLE
prov/gni: Dynamic Process Connection

### ISSUE THREAD

**Issue by JoZie** [ofiwg/libfabric#4153]
I want to connect 2 MPI applications. Therefor I wrote a libfabric communication library. For the "Job Connection" feature I did the initial implementation on my local machine with the `sockets` provider. This works just fine. But when I'm switching to our Cray HPC system with the `gni` provider the processes of the two aprun instances won't connect.
The log for one of the processes (0) states that there is a connection attempt. But there is on response from the other process (1). 

Is is possible to connect 2 jobs using libfabric with the `gni` provider? And Do the jobs have to share the same `auth_key`?

**Process 0 log:**
```
...
libfabric:gni:av:gnix_av_lookup():565<trace> [25647:1]
libfabric:gni:av:_gnix_av_lookup():495<trace> [25647:1]
I'm:  0001:0x000018ec:0x0042f001:01:0x42f001:0x00000000:00
libfabric:gni:av:gnix_av_lookup():565<trace> [25647:1]
libfabric:gni:av:_gnix_av_lookup():495<trace> [25647:1]
send to:  0001:0x000018ee:0x008cd001:01:0x8cd001:0x00000000:00
libfabric:gni:core:_gnix_fr_alloc():242<debug> [25647:1] 0x658578 refs 2
libfabric:gni:ep_data:_gnix_send():3203<debug> [25647:1] Queuing (0x7fffffff5ecc 4)
libfabric:gni:ep_ctrl:_gnix_vc_ep_get_vc():2157<trace> [25647:1]
libfabric:gni:ep_ctrl:__gnix_vc_get_vc_by_fi_addr():218<trace> [25647:1]
libfabric:gni:ep_ctrl:__gnix_vc_get_vc_by_fi_addr():222<debug> [25647:1] ep->vc_table = 0x69c570, ep->vc_table->vector = 0x69c5e0
libfabric:gni:ep_ctrl:_gnix_vec_at():220<trace> [25647:1]
libfabric:gni:ep_ctrl:__gnix_vec_at():219<debug> [25647:1] There is no element at index (1) in __gnix_vec_at
libfabric:gni:av:_gnix_av_lookup():495<trace> [25647:1]
libfabric:gni:av:_gnix_av_lookup():495<trace> [25647:1]
libfabric:gni:ep_ctrl:_gnix_vc_alloc():1453<trace> [25647:1]
libfabric:gni:ep_ctrl:_gnix_nic_get_rem_id():650<trace> [25647:1]
libfabric:gni:ep_ctrl:__gnix_vec_insert_at():151<trace> [25647:1]
libfabric:gni:ep_ctrl:_gnix_vc_connect():1657<trace> [25647:1]
libfabric:gni:ep_data:_gnix_vc_queue_tx_req():1954<debug> [25647:1] Queued request (0x658e88) on busy VC
libfabric:gni:ep_ctrl:_gnix_vc_schedule():2115<debug> [25647:1] Scheduled VC (0x6820c8)
libfabric:gni:ep_ctrl:__gnix_vc_conn_req_prog_fn():1303<trace> [25647:1]
libfabric:gni:ep_ctrl:_gnix_mbox_alloc():695<trace> [25647:1]
libfabric:gni:ep_ctrl:__page_count():246<debug> [25647:1] Mbox_size: 66432, mpmmap: 2048, page_size: 2097152
libfabric:gni:ep_ctrl:__page_count():250<debug> [25647:1] Total size: 136314880, page_count: 65
libfabric:gni:ep_ctrl:__gnix_vc_conn_req_prog_fn():1369<debug> [25647:1] conn req tx: (From Aries addr 0x18ec Id 4386817 to Aries 0x18ee Id 9228289 CM NIC Id 9228289 vc 0x6820c8)
libfabric:gni:ep_ctrl:_gnix_xpmem_get_my_segid():511<trace> [25647:1]
libfabric:gni:ep_ctrl:_gnix_cm_nic_send():431<trace> [25647:1]
libfabric:gni:ep_ctrl:_gnix_dgram_alloc():244<trace> [25647:1]
libfabric:gni:ep_ctrl:_gnix_dgram_bnd_post():341<trace> [25647:1]
libfabric:gni:ep_ctrl:__gnix_vc_conn_req_prog_fn():1405<debug> [25647:1] moving vc 0x6820c8 state to connecting
libfabric:gni:ep_ctrl:__gnix_nic_next_pending_vc():2046<info> [25647:1] Dequeued progress VC (0x6820c8)
```

**Process 1 log:**
```
...
libfabric:gni:av:gnix_av_lookup():565<trace> [26829:1]
libfabric:gni:av:_gnix_av_lookup():495<trace> [26829:1]
libfabric:gni:av:gnix_av_lookup():565<trace> [26829:1]
I'm:  0001:0x000018ee:0x008cd001:01:0x8cd001:0x00000000:00
Recv from:  0001:0x000018ec:0x0042f001:01:0x42f001:0x00000000:00
libfabric:gni:av:_gnix_av_lookup():495<trace> [26829:1]
libfabric:gni:ep_data:_gnix_recv():2614<debug> [26829:1] posted_queue = 0x658418
libfabric:gni:ep_ctrl:__remove_by_tag_and_addr():872<debug> [26829:1] removing a message by tag, ts=0x6583b8 tag=1000000000000 ignore=7000000000000000 flags=0 context=0x6a09f0 addr=0x7fffffff5c10                                                                             
libfabric:gni:ep_ctrl:__remove_by_tag_and_addr():874<debug> [26829:1] ret=(nil)
libfabric:gni:core:_gnix_fr_alloc():242<debug> [26829:1] 0x658578 refs 2
libfabric:gni:ep_data:_gnix_recv():2763<debug> [26829:1] New recv, req: 0x658e88
libfabric:gni:ep_ctrl:_gnix_insert_tag():842<debug> [26829:1] inserting a message by tag, ts=0x658418 tag=1000000000000 req=0x658e88
libfabric:gni:ep_ctrl:_gnix_insert_tag():850<debug> [26829:1] ret=0
```

**Comment by jswaro** [comment]
The jobs must have the same auth key. If this is a two distinct job scenario, they must share a common auth key. If this is within the same job, they must share a common auth key OR the default auth key.

**Comment by jswaro** [comment]
So, when trying to use the GNI provider between two different jobs, DRC is necessary for the allocation of authorization keys. 

See: https://cug.org/proceedings/cug2016_proceedings/includes/files/pap108s2-file1.pdf

The authorization key must be built by the user application, but one of the elements of the fi_gni_auth_key is a protection key which you can read from the DRC credentials. drc_get_first_cookie() is the function call that can be used to do that, but mind you that requires building with libDRC to do that. 

https://github.com/ofiwg/libfabric/blob/master/prov/gni/include/fi_ext_gni.h

**Comment by JoZie** [comment]
Thank you! I managed to create an auth_key with `apmgr pdomain ...`.

To get the connection finally working I had to tweak the calls to `_gnix_auth_key_lookup` within `__create_slab` and `__nic_setup_irq_cq`. Those calls still take `GNIX_PROV_DEFAULT_AUTH_KEY` as arguments, which will not work with a user defined key. I just took the cookie from the `gnix_nic` struct and created a key.

I will open a PR for the patch, but before I have to tweak the handling of the default auth key, especially the key of the hash table.
